### DISCUSSION TITLE
`cus` 项目中的一些疑问

### DISCUSSION THREAD

**Discussion by xkwxdyy** [Sophanatprime/cus#37] [Ideas]
## 说明

我以一位普通用户的角度在看 `cus-cn.pdf`，然后希望能够提出一些优化（降低读者阅读门槛）的角度，主要是一些希望增加、补充的注释或者内容。

虽然用户要熟悉 `cus` 的内容，自己肯定也需要上手敲，但是个人还是希望不要写得太 API，`interface3.pdf` 那种真的很劝退，所以开启此 discussion 希望能让手册对一般用户友好点，

**Comment by xkwxdyy** [comment]
## 关于 `makelapbox`

https://github.com/Sophanatprime/cus/blob/aa23963ee748c79f5a481fbf491532c3dea03fbd/cus-cn.tex#L458-L471

这里没说明 `<pos>` 的可选值，可能是考虑到 `\makebox` 是基础命令？但是对大部分用户而言，其实 `\makebox` 也很少用 `pos`，所以这里可以再解释一下 `<pos>` 的可选值有哪些，并且给一些例子。最好也能给例子说明和 `\makebox` 的不同体现在哪，不然我看了一下只知道多了一个 `<lap to> `，但也不知道啥用

**Reply by xkwxdyy** [reply]
```latex
左边的文字\makelapbox[3em][c][r]{测试}右边的文字

左边的文字\makelapbox[3em][l][r]{测试}右边的文字

左边的文字\makelapbox[3em][r][r]{测试}右边的文字

左边的文字\makelapbox[3em][l]{测试}右边的文字
```
但从这个例子可以看到效果了，相当于 `lap to` 是 `pos` 之后做的二次调整。

**Reply by xkwxdyy** [reply]
`\makelapbox` 其实是 `\makebox` 的优化。它增加了 `\makebox` 没有的 `⟨lap to⟩` 可选参数来进一步控制 `⟨material⟩` 的位置。

可以把 `⟨lap to⟩` 看作 `⟨material⟩` 位置的二次调整：
1. `⟨material⟩` 先根据 `⟨pos⟩` 参数调整 `⟨width⟩` 宽的盒子的整体位置；
2. 盒子的位置确定了之后再在 `⟨lap to⟩` 的作用下，调整 `⟨material⟩` 在 `⟨width⟩` 宽的盒子里的位置

**Reply by Sophanatprime** [reply]
dev 版补充了说明，并且修复了相关的 bug。

**Comment by xkwxdyy** [comment]
## `enumlist` 环境

```latex
\begin{enumlist} [⟨default label⟩] {⟨left⟩} {⟨indent⟩} {⟨label sep⟩} {⟨right⟩} ...
\end{enumlist}
```

请问 `enumlist` 环境的 `⟨default label⟩` 该怎么写呢? 我觉得可以给几个例子，比如 `enumerate`， `itemize` 环境是怎么可以通过 `enumlist` “复刻”出来的。

还有一个问题是 `enumlist` 环境有什么别的应用场景吗

**Reply by Sophanatprime** [reply]
它最开始是用在自定义目录中的，原来没有模板目录的时候，用列表实现缩进比较简洁，大多数只要控制这几个值，所以单独定义了一个环境。

**Reply by xkwxdyy** [reply]
原来是这样，了解了。

**Comment by xkwxdyy** [comment]
## 关于 `\cussetstyle`

我看了一下这个命令的定义以及 `cus` 的一处应用：
https://github.com/Sophanatprime/cus/blob/e497371ac6ce31d2117f9bd8b5f377c668b1a1f2/library/cus.library.doc.tex#L52-L54

但是抱歉我还是不太能理解这个接口的应用场景，可以举一两个例子吗？

**Reply by Sophanatprime** [reply]
`\cussetstyle{key}{value}` 类似于 `\pgfkeys{/cus/key/.style={value}}`；`\cussetstyle*{key}{value}` 类似于 `\pgfkeys{/cus/key/.code={value}}`。只不过是用 `l3keys` 实现的。可选参数是设置所属路径的（手册键值选项旁注的灰色部分）。

**Reply by xkwxdyy** [reply]
抱歉我 `pgf` 用的也很少。我感觉这个有点像 template 的感觉？还是说 `.style` 和 `.code` 之间是没有联系的呢？

> `\cussetstyle{key}{value}` 类似于 `\pgfkeys{/cus/key/.style={value}}`

那这个 `.style` 要怎么使用呢？

**Reply by Sophanatprime** [reply]
v0.0.5b https://github.com/Sophanatprime/cus/commit/4bb96f344b1fa6f16b21cfa733f3a4482985073e 补充了例子。

`.style` `.code` 是用 `pgfkeys` 类比 `\cussetstyle` 和 `\cussetstyle*`。
`.style` 的参数是已有的键，使用 `key` 相当于使用了这些键，效果等同于 `l3keys` 的 `.meta:n`。
`.code` 的参数是一些代码，使用 `key` 相当于使用了这些代码，效果等同于 `l3keys` 的 `.code:n`。

**Comment by xkwxdyy** [comment]
## 关于 `spreadtext`

我看到手册上说 `\speadtext*` 可以在两端加一段间距，但是这个间距好像没有提供接口来改？

另外武汉大学的学位论文封面（word）有两种 spread 的效果想请教一下：

### 名字和职称的分散对齐
![image](https://github.com/Sophanatprime/cus/assets/87134063/57484c39-c0cf-462f-b266-4373d39bdc88)

“教授”两字的间距稍微窄一点，但是和名字之间的间距又比名字本身的间距宽一点，我能想到的就是把“教授”单独用一个 `spreadtext` 然后整体再用一个 `spreadtext`，但是不知道怎么控制名字和“教授”之间的 spread，因为他更长一点，可能要把名字也放在一个 `spreadtext` 里面？

### 换行的分散对齐
![image](https://github.com/Sophanatprime/cus/assets/87134063/66e19d3e-3501-403f-893f-161171b0fc8a)

目前`spreadtext`并没有提供这种换行的 spread 效果，常规的分散比如用 `\hfil` 或者 `makebox` 好像也是一个水平盒子而实现不了这种效果，请问有什么办法吗？

**Reply by xkwxdyy** [reply]
换行的分散对齐，tex.se 上我找到一个[回答](https://tex.stackexchange.com/a/226174/246645)，然后改了一下：
```latex
\documentclass{ctexart}

\makeatletter
\renewcommand{\title}[1]{\gdef\@title{\DoUppercase{#1}}}
\DeclareRobustCommand\DoUppercase[1]{\Do@Uppercase#1\\\@nil}
\def\Do@Uppercase#1\\#2\@nil{%
  % \MakeUppercase{#1}%
  \makebox[10em][s]{#1}%
  \ifx\\#2\\%
    \expandafter\@gobble
  \else
    \\\expandafter\@firstofone
  \fi
  {\Do@Uppercase#2\@nil}%
}
\makeatother

\title{测试一下哦哈哈哈\\一下哦哈}

\begin{document}
\maketitle
\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/d56d1adf-3680-4e83-a6e5-95246ab80e19)

确实符合预期。不知道 cus 能否封装一个类似的命令？

**Reply by xkwxdyy** [reply]
名字的需求，我的那个想法行不通
```latex
\documentclass{ctexart}
\makeatletter
\@ifundefined{input@path}{\let\input@path\@empty}{}
\g@addto@macro\input@path{{./module}{./library}{./lt3ekeys}{./cus-doc}{./cus-doc-aux}}
\makeatother
\usepackage{cus}


\begin{document}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼羊}%
    \spreadtext{3em}{教授}%
  }
}

\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/f7d39f10-1421-4ef9-a7c4-d9d3b360c101)

而且我这样有个问题，如果名字是两个字到四个字不等，“教授”改成“副教授”的话，字数一变，内层的分散的宽度貌似就只能手动改：
```latex
\documentclass{ctexart}
\makeatletter
\@ifundefined{input@path}{\let\input@path\@empty}{}
\g@addto@macro\input@path{{./module}{./library}{./lt3ekeys}{./cus-doc}{./cus-doc-aux}}
\makeatother
\usepackage{cus}


\begin{document}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{3em}{夏大}%
    \spreadtext{2.5em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼}%
    \spreadtext{2.5em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼羊}%
    \spreadtext{4em}{副教授}%
  }
}

\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/9389f441-4b1c-492d-8427-06acbfbaa884)

虽然可以让用户接口为 clist，然后分别对两个 item 进行长度判断，但感觉还是挺麻烦的。word 的分散对齐会进行合适的处理：
![image](https://github.com/Sophanatprime/cus/assets/87134063/c49cab83-656d-4bdb-b6f2-71bf61be9d4b)

**Reply by xkwxdyy** [reply]
如果加一个 `\mbox` 的话中间会产生空格，但不会两端对齐
```latex
\documentclass{ctexart}
\makeatletter
\@ifundefined{input@path}{\let\input@path\@empty}{}
\g@addto@macro\input@path{{./module}{./library}{./lt3ekeys}{./cus-doc}{./cus-doc-aux}}
\makeatother
\usepackage{cus}


\begin{document}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{3em}{夏大}%
    \mbox{}
    \spreadtext{2.5em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼}%
    \mbox{}
    \spreadtext{2.5em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼羊}%
    \mbox{}
    \spreadtext{4em}{副教授}%
  }
}

\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/06e39fd6-9904-408a-a16d-5f288715ac0a)

**Reply by xkwxdyy** [reply]
把 `\mbox` 改成  `\hfill` 就行：
```latex
\documentclass{ctexart}
\makeatletter
\@ifundefined{input@path}{\let\input@path\@empty}{}
\g@addto@macro\input@path{{./module}{./library}{./lt3ekeys}{./cus-doc}{./cus-doc-aux}}
\makeatother
\usepackage{cus}


\begin{document}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{3em}{夏大}%
    \hfill
    \spreadtext{3em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{4em}{夏大鱼}%
    \hfill
    \spreadtext{3em}{教授}%
  }
}

\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{6em}{夏大鱼羊}%
    \hfill
    \spreadtext{4em}{副教授}%
  }
}

\end{document}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/354005ad-95c8-4da0-a6f2-e059225a6635)
但这样的话 `\spreadtext*` 的 `*` 作用就没了

**Reply by Sophanatprime** [reply]
`\spreadtext` 实际上是把单词和 CJK 文字之间的间距修改为 `\hfil`，v0.0.5b https://github.com/Sophanatprime/cus/commit/4bb96f344b1fa6f16b21cfa733f3a4482985073e 新增了更加底层的命令 `\useinfiniteskip`。

> 我看到手册上说 \speadtext* 可以在两端加一段间距，但是这个间距好像没有提供接口来改？

这个间距就是 `\hfil`，v0.0.5b 可以设置几倍 `\hfil`。

> 换行的分散对齐，tex.se 上我找到一个[回答](https://tex.stackexchange.com/a/226174/246645)

这个会导致 underfull，v0.0.5b 补充了类似的但不会有 underfull 的例子。后面的这个需求也在 v0.0.5b 中给了类似的例子。

 ```tex
\fbox{%
  \spreadtext*{6.17cm}{%
    \spreadtext{3em}{夏大}%
    \spreadtext{2.5em}{教授}%
  }
}
```
取消第三行的注释符就行了。加了 `\mbox` 就有效，是因为 `\mbox{}` 后面没有注释符，TeX 自动在 `\mbox` 后面加上一个空格。`\spreadtext` 再把这个空格替换为无限的长度。

> 把 `\mbox` 改成 `\hfill` 就行：

这是因为 `\spreadtext` 本来就是会加上无限的长度，手动加上自然也可以。`*` 无效，是因为 `\hfill` 的阶数太高，换成低阶的 `\hfil` 就行。也可参考 v0.0.5b 手册增加的例子。

**Reply by xkwxdyy** [reply]
感谢，照着手册的例子改了一下，效果很不错。

**Reply by xkwxdyy** [reply]
但是那个换行的那个：
```latex
% cus-cn.tex
\parbox{8\ccwd}{\useinfiniteskip{1fil}%
  \rightskip0pt plus 1fil \parfillskip-\rightskip
  测试一下哦哈哈哈 ac b 哦哈}
```
如果我改成手动换行：
```latex
\parbox{5\ccwd}{\useinfiniteskip{1fil}%
  \rightskip0pt plus 1fil \parfillskip-\rightskip
  测试一下\\测试一下\\ 测试一下测试一下}
```
![image](https://github.com/Sophanatprime/cus/assets/87134063/479d1111-a089-41c1-b3a8-8ddcee5329d2)
会发现断行的几行尾部有多余的距离，没有进行分散对齐，请问这是什么原因呢？以及请问一下 ` \rightskip0pt plus 1fil \parfillskip-\rightskip` 这段代码的作用是？ ` \rightskip0pt plus 1fil` 是把 `\rightskip` 设置为 `0pt plus 1fil`，但不太理解 `\parfillskip-\rightskip` 的作用

**Reply by Sophanatprime** [reply]
> 会发现断行的几行尾部有多余的距离，没有进行分散对齐，请问这是什么原因呢？

这是由于在换行时会添加 `\rightskip`，这里是 `0pt plus 1fil`，它和 `\useinfiniteskip` 设置的间距一样，因此会在每行后面添加相同的间距。

在每段末尾会自动添加 `\parfillskip`，设置它为 `-\rightskip` 是为了和每行添加的 `\rightskip` 相抵消，这样每段最后就没有间距了，就会使用 `\useinfiniteskip` 设置的间距，否则也会和前 2 行一样。

只要换成更高阶的长度就行了。
```
\fbox{\parbox{5\ccwd}{\useinfiniteskip{1fill}
  测试一下\\测试一下\\ 测试一下测试一下}}
```

**Reply by xkwxdyy** [reply]
了解了，感谢！

**Comment by xkwxdyy** [comment]
## 关于 `\setuponetitleq`

我看了一下手册，没有具体解释这个命令和 `\setuponetitle` 的区别，我看了一下源码，应该是 `\setuponetitleq` 的第一个参数不能是数字吧？只能是 `chapter` 这些？`\setuponetitle` 还会检测设置的 title level 是否存在，但是 `\setuponetitleq` 没有这个检测。

这个名字我觉得有误导性，我理解的是 `set` `one` `title` `eq`，但是感觉 `eq` 在这里的语义并不太突出。您觉得呢？

这样看，我不太知道 `\setuponetitleq` 的作用是什么，有 `\setuponetitle` 了为什么还需要 `\setuponetitleq` 呢

**Reply by Sophanatprime** [reply]
`\setuponetitleq` 是 `set` `one` `title` `q`(quick)。

`\setuponetitle` 确实包含了 `\setuponetitleq` 的功能，正常情况用前者就好了，后者不检查 `⟨title⟩` 是否为整数，也不检查存在性，稍微快一点点（不过差距也很小）。

**Reply by xkwxdyy** [reply]
原来如此，我以为是因为  `title` 和 `eq` 的 e 重合了，所以只取一个，了解了，感谢。建议在手册里提一下。

**Comment by xkwxdyy** [comment]
## 关于 `bool` 类型的键值的默认值问题

我在阅读手册的时候发现很多 `bool` 键值
![image](https://github.com/Sophanatprime/cus/assets/87134063/fe156c0f-956d-49c0-ad05-aee7629f28a4)

会采用 `&` 的语法糖来写初始值，但是同时前面又用了 `\TTF` 这种命令来表示初始值，而且经常出现和 `&` 的初始值不同的情况，这是有意的还是您忘记改了？因为数量比较多，且我不确定您是不是出于什么考虑的，我就没提 pr 了，先问问。

**Reply by Sophanatprime** [reply]
`&` 表示“可选值”，里面加粗的是“默认值”，后面蓝色的是“初始值”。

**Reply by xkwxdyy** [reply]
1. `\TTF` 系列的命令肯定是可选，要不要直接从源上改一下效果？这样这些命令前就不用 `&` 了？
2. 默认值是指 `defalt` 那种吗？比如 `hang` 就表示 `hang=true`? 这样的话那我理解了，只是这个可能需要在手册前面说明一下（我没印象有在 `cus-cn` 里看到）

**Comment by xkwxdyy** [comment]
## 关于 `\cus_act_case_true:nnnn` 的例子

https://github.com/Sophanatprime/cus/blob/28913e60ef59a4930e8e7f440a95c909ec95094b/cus-cn.tex#L3929-L3944

请问
- `tl  ~ \use:n` 为什么要加 `\use:n`
- `not ~ equal \use_none:n` 为什么要加 `\use_none:n`
呢？

**Reply by Sophanatprime** [reply]
`\cus_act_case_true:nnnn` 的第四个参数是参数列表，它会附加到第二个参数的偶数项和第三个参数的后面，相当于
```
\exp_args:No \cus_act_case_true:nnnn  
  { \token_to_str:N : } % \catcode`\:=12 
  {  
    { \tl_if_eq:nnTF  { : }       } { tl  ~ \use:n {eq} } % \catcode`\:=11 不相等 
    { \token_if_eq_catcode:NNTF : } { cat ~ \use:n {eq} } % \catcode`\:=11 不相等 
    { \str_if_eq:nnTF { : }       } { str ~ \use:n {eq} } % string 相等 
  }
  { not ~ equal \use_none:n {eq} }
  { }
```
它适用于多个同类且连续的判断，避免嵌套。

**Reply by xkwxdyy** [reply]
` \use:n {eq}` 和直接的 `eq` 一样吗？也就是上面的 `\use:n` 能否去掉呢

**Reply by Sophanatprime** [reply]
`\use:n {eq}` 展开为 `eq`，如果去掉 `\use:n` 则 `eq` 外面会多出一对花括号。

**Reply by xkwxdyy** [reply]
大概明白了，应该是因为 `{ not ~ equal \use_none:n }` 吧？就是不想在这个地方出现 `eq`, 所以用了  `\use_none:n` 和 `{eq}`，然而上面如果不用 `\use:n` 的话，`{eq}` 就会多一对花括号

**Comment by xkwxdyy** [comment]
## 关于 `\cus_hook_gput_pe:n`

我在 `index` 库中发现这个钩子，但是我在 cus 项目中直接搜索并没有搜索到，请问这个钩子的作用是？

原始需求：是想把 index 处理移植到一个自己写的一个文类里，但是复制后报错，说 undefined。

如果要单独调用 index 库的话，有什么依赖吗

**Reply by xkwxdyy** [reply]
我找到定义的地方了：https://github.com/Sophanatprime/cus/blob/e0857c5abec658a79927240e05de52122fe4bfbc/cus.sty#L467-L488

但是这个貌似依赖蛮多变量的。

如果我只想单独使用 index 库的话，感觉把 `\cus_hook_gput_pe:n` 改成 `\AtEndPreamble` 是不是可以？

**Reply by xkwxdyy** [reply]
`AtEndPreamble` 不行，改成 `\AddToHook { begindocument / before }` 就行了，我再测试一下 index 功能能否使用

**Reply by xkwxdyy** [reply]
经过下面的修改后可以：
1. 添加 `framedmulticol` 宏包
2. 添加https://github.com/Sophanatprime/cus/blob/e0857c5abec658a79927240e05de52122fe4bfbc/module/cus.module.util.tex#L750-L753
3. https://github.com/Sophanatprime/cus/blob/e0857c5abec658a79927240e05de52122fe4bfbc/module/cus.module.index.tex#L108 改为 `\exp_not:N \AddToHook { enddocument/afteraux }`

**Reply by Sophanatprime** [reply]
`cus` 的模块和库的耦合程度高，设计之初就没有（也不会）考虑独立于 `cus` 使用。如果要单独使用某个模块或库的话，肯定不容易。要在其它文档类中使用类似的功能，也可以考虑修改其它已有的宏包。比如 `imakeidx` 就可以制作索引。

**Reply by xkwxdyy** [reply]
好的，感谢
